controller/volume_controller.go

```diff
@@ -1125,77 +1182,401 @@ func (vc *VolumeController) ReconcileVolumeState(v *longhorn.Volume, es map[stri
+		}
+
+		// TODO: reconcileVolumeSize
+		// The engine expansion is complete
+		if v.Status.ExpansionRequired && v.Spec.Size == e.Status.CurrentSize {
+			v.Status.ExpansionRequired = false
+			v.Status.FrontendDisabled = false
+		}
+	}
+
+	if err := vc.checkAndFinishVolumeRestore(v, e, rs); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (vc *VolumeController) reconcileAttachDetachStateMachine(v *longhorn.Volume, e *longhorn.Engine, rs map[string]*longhorn.Replica, isNewVolume bool, log *logrus.Entry) error {
+	//TODO: link the state machine graph here
+
+	if isNewVolume || v.Status.State == "" {
+		v.Status.State = longhorn.VolumeStateCreating
+		return nil
+	}
+
+	if v.Spec.NodeID == "" {
+		if v.Status.CurrentNodeID == "" {
+			switch v.Status.State {
+			case longhorn.VolumeStateAttached, longhorn.VolumeStateAttaching, longhorn.VolumeStateCreating:
+				vc.closeVolumeDependentResources(v, e, rs)
+				v.Status.State = longhorn.VolumeStateDetaching
+			case longhorn.VolumeStateDetaching:
+				vc.closeVolumeDependentResources(v, e, rs)
+				if vc.verifyVolumeDependentResourcesClosed(e, rs) {
+					v.Status.State = longhorn.VolumeStateDetached
+					vc.eventRecorder.Eventf(v, v1.EventTypeNormal, constant.EventReasonDetached, "volume %v has been detached", v.Name)
+				}
+			case longhorn.VolumeStateDetached:
+				// This is a stable state.
+				// We attempt to close the resources anyway to make sure that they are closed
+				vc.closeVolumeDependentResources(v, e, rs)
+			}
+			return nil
+		}
+		if v.Status.CurrentNodeID != "" {
+			switch v.Status.State {
+			case longhorn.VolumeStateAttaching, longhorn.VolumeStateAttached, longhorn.VolumeStateDetached:
+				vc.closeVolumeDependentResources(v, e, rs)
+				v.Status.State = longhorn.VolumeStateDetaching
+			case longhorn.VolumeStateDetaching:
+				vc.closeVolumeDependentResources(v, e, rs)
+				if vc.verifyVolumeDependentResourcesClosed(e, rs) {
+					v.Status.CurrentNodeID = ""
+					v.Status.State = longhorn.VolumeStateDetached
+					vc.eventRecorder.Eventf(v, v1.EventTypeNormal, constant.EventReasonDetached, "volume %v has been detached", v.Name)
+				}
+			}
+			return nil
+		}
+	}
+
+	if v.Spec.NodeID != "" {
+		if v.Status.CurrentNodeID == "" {
+			switch v.Status.State {
+			case longhorn.VolumeStateAttached:
+				vc.closeVolumeDependentResources(v, e, rs)
+				v.Status.State = longhorn.VolumeStateDetaching
+			case longhorn.VolumeStateDetaching:
+				vc.closeVolumeDependentResources(v, e, rs)
+				if vc.verifyVolumeDependentResourcesClosed(e, rs) {
+					v.Status.State = longhorn.VolumeStateDetached
+					vc.eventRecorder.Eventf(v, v1.EventTypeNormal, constant.EventReasonDetached, "volume %v has been detached", v.Name)
+				}
+			case longhorn.VolumeStateDetached:
+				if err := vc.openVolumeDependentResources(v, e, rs, log); err != nil {
+					return err
+				}
+				v.Status.State = longhorn.VolumeStateAttaching
```

Comment:
Similar question as above

Reply:
similar to https://github.com/longhorn/longhorn-manager/pull/1541#discussion_r1120953899